test/test_ops.py

```diff
@@ -227,17 +227,26 @@ def _ref_test_helper(self, ctx, device, dtype, op, skip_zero_numel=False, skip_z
                 )
             ):
                 continue
+
+            # Note: this deliberately uses 'op.torch_opinfo' in both cases
+            # because 'op.aliases' is empty.  When the aliases are executed
+            # within the context, they will behave as refs.  This is done to
+            # check for compatibility between all APIs.
+            aliases = op.torch_opinfo.aliases if op.test_aliases else ()
             with ctx():
-                ref_result = op(sample.input, *sample.args, **sample.kwargs)
-            torch_result = op.torch_opinfo(sample.input, *sample.args, **sample.kwargs)
+                ref_results = [f(sample.input, *sample.args, **sample.kwargs)
+                               for f in itertools.chain((op,), aliases)]
+            torch_results = [f(sample.input, *sample.args, **sample.kwargs)
+                             for f in itertools.chain((op.torch_opinfo,), aliases)]
```

Comment:
How does this work? Isn't the `chaining` done with the same `aliases` in both cases, meaning the the same thing is being compared?

Reply:
thanks for bringing this up, there are multiple issues here. apologies for the long comment, i've tried to explain it as brief as i can.

## about the context

for instance, with `TorchRefsMode(strict=True)` as `ctx`, a ref with `torch.foo` in it will call `_refs.foo` instead (this is already documented in `test_python_ref`). more details below.

## testing the op against the ref

let's first look at how we test the op against its ref. when we test `op` against `op.torch_opinfo` here, we call `_refs.foo` (that's `op`, but it's in the context, so all inner `torch` calls in the ref function body are mapped to ref impls as described above) against `torch.foo` (that's `op.torch_opinfo`, but it's outside of the context, so execution continues as in regular pytorch).

this is a good test because we check different entrypoints, but i now realize it's also not great because users will never call `_refs` functions directly. iiuc, they will just write regular pytorch code inside a new context, maybe like this:

```py
with primtorch():  # made-up future context for primtorch
  torch.log_softmax(...)
  # the rest of the code that should be copy-pastable from regular pytorch
```

_i think we need to add another pair of input functions where we check `op.torch_opinfo` against itself inside and outside the context_ because that's what users will actually be executing (if my understanding is correct).

## testing aliases

with aliases, we just call stuff in `torch` in both `ref_results` and `torch_results`. but in `ref_results`, inner calls to `torch` are translated to `_refs` calls instead (due to it being in the context).

for instance, we might have `torch.nn.functional.log_softmax` as an alias in both cases:

```py
def log_softmax(input: Tensor, dim: Optional[int] = None, _stacklevel: int = 3, dtype: Optional[DType] = None) -> Tensor:
    if has_torch_function_unary(input):
        # [1]
        return handle_torch_function(log_softmax, (input,), input, dim=dim, _stacklevel=_stacklevel, dtype=dtype)
[...]
    return ret
```

in the context, the `log_softmax` call at `[1]` will go to `_refs.nn.functional.log_softmax`.

_perhaps we could try adding another set of inputs where we'd test `_refs.alias` against `torch.alias`, but we have no convenient way of getting to `_refs.alias` as of now. and based on the assumption above, those will never be executed directly by users, so it's seems not worth it._

## summary

rather than doing anything about the aliases, i'm thinking about adding a test where i test `op.torch_opinfo` against itself (as described above) in addition to the existing tests.

thoughts?

Reply:
i also forgot to mention that most aliases should be defined as `foo = torch.foo` anyway, so it shouldn't matter much in practice(?). there might be some corner cases, like where we have slightly more complicated wrappers for `nn.functional` and `special` variants due to signature incompatibilities, but i don't feel like this justifies complicating this test. and again, the entrypoints that we need to be testing are `torch.foo` (inside and outside the context), iiuc

Reply:
Yes we need to make sure the testing is correctly remapping ops, and that's the most important thing. Maybe we should have a separate test that just checks that the remapping is correct? 

For aliases, there are "true" aliases, which we should define as `foo = torch.foo`, and "near" aliases, as you're pointing out. I'm not that worried about testing either except that it'd be nice to verify both call the aliased operation properly. If that's too tricky to figure out then let's not worry about it for now, it should be easy to implement later.

Reply:
rewrote the comment and introduced new test modes